syntax="proto3";
option cc_generic_services = true;

package storage_service;

message LogWriteRequest {
    bytes log = 1;
    message PageID{
    	bytes table_name = 1;
        sint32 page_no = 2;
    }
    repeated PageID page_id = 2;
};

message RaftLogWriteRequest{
    bytes raft_log = 1;
}

message LogWriteResponse{}

message GetPageRequest {
	message PageID{
    	bytes table_name = 1;
        sint32 page_no = 2;
    }
	repeated PageID page_id = 1;
    uint64 require_batch_id = 2;
};

message GetPageResponse {
    bytes data = 1;
};

message GetBatchIndexRequest {
	bytes table_name = 1;
    uint64 batch_id = 2;
};

message GetBatchIndexResponse {
    repeated uint64 itemkey = 1;
    repeated sint32 pageid = 2;
    repeated sint32 slotid = 3;
};

service StorageService {
    rpc LogWrite(LogWriteRequest) returns (LogWriteResponse);
    rpc RaftLogWrite(RaftLogWriteRequest) returns (LogWriteResponse);
    rpc GetPage(GetPageRequest) returns (GetPageResponse);
    rpc PrefetchIndex(GetBatchIndexRequest) returns (GetBatchIndexResponse);
};
